crates/subspace-wasm-tools/src/lib.rs

```diff
@@ -46,13 +47,32 @@ pub fn create_runtime_bundle_inclusion_file(
 
     let execution_wasm_bundle_rs_path =
         env::var("OUT_DIR").expect("Set by cargo; qed") + "/" + target_file_name;
-    let execution_wasm_bundle_rs_contents = format!(
+    let mut execution_wasm_bundle_rs_contents = format!(
         r#"
             pub const {bundle_const_name}: &[u8] = include_bytes!("{}");
         "#,
         execution_wasm_bundle_path.escape_default()
     );
 
+    if let Some(link_section_name) = maybe_link_section_name {
+        let wasm_bundle_content = std::fs::read(&execution_wasm_bundle_path).unwrap_or_else(|e| {
+            panic!("Failed to read wasm bundle from {execution_wasm_bundle_path:?}: {e}",)
+        });
+        let wasm_bundle_size = wasm_bundle_content.len();
+
+        let link_section_decl = format!(
+            r#"
+        const _: () = {{
+            #[cfg(not(feature = "std"))]
+            #[link_section = "{link_section_name}"]
+            static SECTION_CONTENTS: [u8; {wasm_bundle_size}] = {wasm_bundle_content:?};
```

Comment:
I'm afraid this is not the way to add data into the file. Use `include_bytes!()` so that compiler can include it in whatever way it wants.

Also I feel like `SECTION_CONTENTS` can cause conflicts, or is that why you had `const` stuff above?

Reply:
Had the same thought, but `include_bytes!()` won't work because it returns a reference but `link_section` can't take a reference as the compiler said.

Yeah, `const _: () = {{` is necessary to avoid conflicts when we embed multiple runtimes into the system domain runtime, this also answer the comment above.

Reply:
I believe you should be able to do `try_into()` in const environment though.

Reply:
```diff
diff --git a/crates/subspace-wasm-tools/src/lib.rs b/crates/subspace-wasm-tools/src/lib.rs
index 11ef8ab20..aa744de02 100644
--- a/crates/subspace-wasm-tools/src/lib.rs
+++ b/crates/subspace-wasm-tools/src/lib.rs
@@ -65,9 +65,10 @@ pub fn create_runtime_bundle_inclusion_file(
                 const _: () = {{
                     #[cfg(not(feature = "std"))]
                     #[link_section = "{link_section_name}"]
-                    static SECTION_CONTENTS: [u8; {wasm_bundle_size}] = {wasm_bundle_content:?};
+                    static SECTION_CONTENTS: [u8; {wasm_bundle_size}] = <[u8; {wasm_bundle_size}]>::try_from(include_bytes!("{}").as_slice()).unwrap();
                 }};
             "#,
+            execution_wasm_bundle_path.escape_default()
         );

         execution_wasm_bundle_rs_contents.push_str(&link_section_decl);
```
Failed to make compiler happy :(

```
  error[E0015]: cannot call non-const fn `<[u8; 879369] as TryFrom<&[u8]>>::try_from` in statics
   --> /mnt/Data2/src/github.com/subspace/subspace/target/release/wbuild/domain-test-runtime/target/wasm32-unknown-unknown/release/build/domain-test-runtime-ae497eff5cbdc16a/out/core_payments_wasm_bundle.rs:7:61
    |
  7 | ... = <[u8; 879369]>::try_from(include_bytes!("/mnt/Data2/src/github.com/subspace/subspace/target/release/wbuild/core-payments-domain-runtime/core_payments_domain_runtime.compact.wasm").as_slice()).un...
    |       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: calls in statics are limited to constant functions, tuple structs and tuple variants
    = note: consider wrapping this expression in `Lazy::new(|| ...)` from the `once_cell` crate: https://crates.io/crates/once_cell

  error[E0015]: cannot call non-const fn `Result::<[u8; 879369], TryFromSliceError>::unwrap` in statics
   --> /mnt/Data2/src/github.com/subspace/subspace/target/release/wbuild/domain-test-runtime/target/wasm32-unknown-unknown/release/build/domain-test-runtime-ae497eff5cbdc16a/out/core_payments_wasm_bundle.rs:7:253
    |
  7 | ...untime/core_payments_domain_runtime.compact.wasm").as_slice()).unwrap();
    |                                                                   ^^^^^^^^
    |
    = note: calls in statics are limited to constant functions, tuple structs and tuple variants
    = note: consider wrapping this expression in `Lazy::new(|| ...)` from the `once_cell` crate: https://crates.io/crates/once_cell

  For more information about this error, try `rustc --explain E0015`.
  error: could not compile `domain-test-runtime` due to 2 previous errors
warning: build failed, waiting for other jobs to finish...
error: failed to run custom build command for `system-domain-runtime v0.1.0 (/mnt/Data2/src/github.com/subspace/subspace/domains/runtime/system)`
```

Reply:
The fix was simple, just pushed it.